frontend-utils/src/parse.rs

```diff
@@ -88,18 +88,46 @@ impl<T: fmt::Debug> fmt::Debug for ParseDetails<T> {
 }
 
 impl<T> ParseDetails<T> {
-    pub fn add_warning(&mut self, message: String) {
-        self.warnings.push(message);
-    }
-
     pub fn values(&self) -> &T {
         &self.result
     }
 }
 
+#[derive(Debug, PartialEq, Eq)]
+pub enum ParseWarning {
+    InvalidToml(TomlError),
+    UnexpectedType {
+        expected: &'static str,
+        actual: &'static str,
+        path: String,
+    },
+    UnsupportedValue {
+        value: String,
+        path: String,
+    },
+}
+
+impl ParseWarning {
+    /// Creates a human readable string from the warning.
+    pub fn display(&self) -> String {
```

Comment:
Why not impl std::fmt::Display?

Reply:
Mostly because if we do want to translate these in the future, its not trivial to pass a language identifier into `std::fmt::Display`, though we could keep `Display` as a wrapper defaulting to `en-US`.

Reply:
Changed to fmt::Display for now. I'll let someone else figure out the way to translate these :smile: